### ISSUE TITLE
`__extenddftf2` is found to be missing at runtime when Rust's `cargo` is driving the build

### ISSUE THREAD

**Issue by adamnovak** [WebAssembly/wasi-sdk#361]
I'm trying to build `rusqlite` for Rust, using `wasi-sdk` 20 as my `CC_wasm32_wasi` environment variable value.

This appears to work, but at runtime in `wasmtime` I can't load the resulting binaries, because `__extenddftf2` is missing.

```
wasmtime target/wasm32-wasi/release/query.wasm --sample "_gbwt_ref" --contig x --interval 1..10 --context 100 --distinct x.gbz.db
Error: failed to run main module `target/wasm32-wasi/release/query.wasm`

Caused by:
    0: failed to instantiate "target/wasm32-wasi/release/query.wasm"
    1: unknown import: `env::__extenddftf2` has not been defined
```

I noticed that symbol name in `wasi-sdk-20.0/share/wasi-sysroot/share/wasm32-wasi/undefined-symbols.txt`. Isn't it supposed to be defined by the SDK, if the compiler doesn't throw an error when code tries to use it?

Is this builtin somehow out of scope for the SDK? Am I supposed to be using another library to provide it? Do I need to somehow convince `rusqlite` or SQLite itself that it isn't actually available for its C code?

**Comment by adamnovak** [comment]
It looks like this symbol is asked for if the C code being compiled tries to use `long double`. Setting `LIBSQLITE3_FLAGS="-DLONGDOUBLE_TYPE=double"` in my `rusqlite` build worked around the problem for me specifically. But it would be nice to have implementations for the `long double` functions introduced by the SDK's compiler.

Maybe the SDK's compiler just shouldn't provide a distinct `long double` if the backing functions aren't written yet? I think `long double` == `double` is an allowed C compiler setup, but `long double` -> program doesn't link probably isn't.

**Comment by sbc100** [comment]
`__extenddftf2` should be defined in compiler-rt (llvm-project/compiler-rt/lib/builtins/extenddftf2.c).   Do you know where you compiler-rt library is coming from?  Can you check it for that symbol?

The version of compiler-rt that ships with wasm-sdk (https://github.com/WebAssembly/wasi-sdk/releases/download/wasi-sdk-20/libclang_rt.builtins-wasm32-wasi-20.0.tar.gz) does seem to define it:

```
$ llvm-nm lib/wasi/libclang_rt.builtins-wasm32.a |  grep __extenddftf2
00000001 T __extenddftf2
```

**Comment by sbc100** [comment]
Yes I see `--allow-undefined` in the link flags.   That seems unfortunate since it delays these types of errors until runtime.

**Comment by sbc100** [comment]
So this seems like a rust issue and not a wasi-sdk issue.  Maybe open an issue there?

**Comment by adamnovak** [comment]
I went ahead and reported this against the Rust compiler builtins library, but I think they are probably going to say it is out of scope for them. If you look at https://github.com/rust-lang/compiler-builtins#unimplemented-functions it lists `extenddftf2.c`, presumably the home of `__extenddftf2`, as explicitly unimplemented, because Rust has no such thing as an `f128` type and thus Rust code never needs to call it.

**Comment by adamnovak** [comment]
They also suggest a way to get a Rust project to use the C compiler's versions of the intrinsics, but it seems to be impossible to do on any stable release of the Rust compiler ever, which presumably means people can't genuinely use it? https://github.com/rust-lang/compiler-builtins/issues/562

**Comment by sbc100** [comment]
In the short term you might also want to just go with the `LIBSQLITE3_FLAGS="-DLONGDOUBLE_TYPE=double"` solution since using 128bit long double is expensive in terms of both code size and performance (in both cases due to the emulation functions).

**Comment by adamnovak** [comment]
I think that is the workaround I'm going to have to take for now. But I'm worried that whatever the next problem is to come out of using the "wrong" intrinsics won't be so easy to bypass.

**Comment by sbc100** [comment]
I agree it seems wrong that rust doesn't include these functions in its compiler-rt in this case.  However I does seem like a rust toolchain issue, not a wasi-sdk one, so perhaps we could close this in favor of an issue in a rust repo?

**Comment by alexcrichton** [comment]
Thanks for the cc, I can try to help out a bit here! Agreed with all of the diagnosis and commentary so far, so I'll offer some thoughts on where this symbol might live.

As y'all have found the Rust version of compiler-builtins lives [here](https://github.com/rust-lang/compiler-builtins) and is a `libcompiler-rt.a` lookalike. This is more-or-less a Rust port of `compiler-rt` from LLVM, and its compilation is tailored for each target (the [`build.rs`](https://github.com/rust-lang/compiler-builtins/blob/master/build.rs) is relatively complicated) and is a mixture of Rust code and C code depending on the target (some all C, some all Rust, some a mix). 

Historically the compiler-builtins crate was intended to be Rust-specific meaning that it contains intrinsics that LLVM might emit calls to for rustc-generated code. This means that historically support for `f128` would not be included because Rust code would never use it. The theory being that if C code needed it then authors of crates could specify `-l/path/to/libcompiler-rt.a` for their system compiler.

Now that being said lots of time has passed in the interim since I worked on things here and so perspectives may have shifted and/or things changed in the meantime. In that sense I'd say that historically intrinsics like this were the responsibility of Rust crates using C, but nowadays it may be shifted elsewhere. I also think that the theory of using the system-provided compiler-rt is probably a disproportionately larger task than might otherwise be required implementing intrinsics like this.

Given all that, I'd recommend one of two routes:

* First confirm upstream whether it's ok to include this symbol or not (e.g. your issue). If perspectives have shifted then it'd be easiest to "just" include this when compiling for wasm.
* Alternatively one option could be to add a Cargo feature to the `compiler-builtins` crate to enable these intrinsics. That way they'd be off-by-default for users of Rust and would only be used for those using the crate in the ecosystem. I'll note though that as you've seen this requires nightly Rust since the crate doesn't compile on stable.

If using Rust stable is a requirement then I'd recommend updating the sqlite build script to pass `-lclang_rt.builtins-wasm32` or similar. Another option would be to distribute that binary with the necessary objects with the sqlite3 crate.

**Comment by adamnovak** [comment]
> If using Rust stable is a requirement then I'd recommend updating the sqlite build script to pass -lclang_rt.builtins-wasm32 or similar. Another option would be to distribute that binary with the necessary objects with the sqlite3 crate.

So this would involve doing something to the `rusqlite` crate so that it exposes to dependent projects that they need to link against the compiler-provided `libcompiler-rt` if they link against `rusqlite`? And then another per-platform environment variable like `CC_wasm32_wasi` to provide the platform-specific `libcompiler-rt` path down to `rusqlite`, so it can then send it back up to dependent crates? (Since I don't think the WASI SDK `cc` implementation has a way to ask for the path from the compiler binary?)

Because I *think* the intrinsics don't get linked in until the very end; I don't think `rusqlite`'s build artifacts are meant to have the intrinsics already linked in and inlined.

**Comment by sbc100** [comment]
Regarding getting the name of libcompiler-rt you can use `clang -print-libgcc-file-name` (verified that this works with wasi-sdk).

**Comment by alexcrichton** [comment]
That would be a localized change to `rusqlite`, yes, and with @sbc100's suggestion you'd execute that and then print out `cargo:rustc-link-search` and `cargo:rustc-link-lib`. It shouldn't require any extra configuration on behalf of users other than already setting `CC_wasm32_wasi`.

**Comment by tgross35** [comment]
Coming by from a chain of linked threads: Rust now has an unstable `f128` type and with it, rust-written versions of all the builtins including `__extenddftf2`. `f16` is unstably supported as well.

**Comment by alexcrichton** [comment]
I think this has since been resolved via a number of possible means, so I'm going to close this. Let me know if this is in error, however.
